jekyll/_posts/2018-10-02-create-a-simple-macro-part-2.md

```diff
@@ -0,0 +1,173 @@
+
+---
+title: "Creating a simple macro 2nd part"
+author: Antonis Geralis
+---
+
+<div class="sidebarblock">
+  <div class="content">
+    <div class="title">Guest post</div>
+    <div class="paragraph">
+      This is a guest post by Antonis Geralis. If you would like to publish 
+      articles as a guest author on nim-lang.org then get in touch with us via
+      <a href="https://twitter.com/nim_lang">Twitter</a> or
+      <a href="https://nim-lang.org/community.html">otherwise</a>.
+    </div>
+  </div>
+</div>
+
+## Creating a macro for a state machine
+
+This time we will be building a state machine. Macros can be useful here in
+order to write cleaner code that is easier to understand. At first we will
+develop a macro that creates a state transition table. Then we will expand on
+this to generate states machines that also run faster.
+
+When designing a macro you will be frequently stopped because `dumpTree`
+can't parse the syntax you hoped for.
+The [manual](https://nim-lang.org/docs/manual.html#macros) explains what types
+of macros are possible. In this example you could be tempted to go for an
+expression-based style like: `let stateTransitions = sparse: ...`. The
+following results in an easier to write macro and maybe more efficient
+generated code:
+
+```nim
+  var stateTransitions: StateTransitionsArray
+
+  sparse(stateTransitions): [
```

Comment:
This macro syntax is not idiomatic Nim. Idiomatic Nim would look like:

```nim
sparse(stateTransitions):
  (Idle, CardInsert): insertCardHandler
  (CardInserted, PinEnter): enterPinHandler
  (PinEntered, OptionSelection): optionSelectionHandler
  (OptionSelected, AmountEnter): enterAmountHandler
  (AmountEntered, AmountDispatch): amountDispatchHandler
```

Reply:
Sure it is valid Nim code, but don't you agree that my suggestion is more readable?